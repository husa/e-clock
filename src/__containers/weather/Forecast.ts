import { connect } from 'react-redux';

import { selectWeatherForecast } from '../../__selectors';

import WeatherForecast from '../../components/weather/Forecast';

const mapStateToProps = (state) => ({
  data: selectWeatherForecast(state),
});

export default connect(mapStateToProps)(WeatherForecast);
