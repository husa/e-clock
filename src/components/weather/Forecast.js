// @flow

import './forecast.styl';

import * as React from 'react';

import type {WeatherResponseForecast} from '../../types';
import lang from '../../services/lang';
import Icon from './Icon';

const isToday = date => (new Date(date)).getDate() === (new Date).getDate();
const isTomorrow = date => (new Date(date)).getDay() === ((new Date).getDay() + 1) % 7;

const getDay = date => {
  if (isToday(date)) return lang.t('Today');
  if (isTomorrow(date)) return lang.t('Tomorrow');
  return lang.t(`Day${(new Date(date)).getDay()}`);
};

type Props = {
  data: WeatherResponseForecast
};

const WeatherForecast = (props: Props) => (
  <div className="weather-forecast">
    {props.data.map(day => (
      <div className="weather-forecast__day" key={day.date}>
        <div className="weather-forecast__day__name">
          {getDay(day.date)}
        </div>
        <Icon
          className="weather-forecast__day__icon"
          code={day.code}
          title={day.text} />
        <div className="weather-forecast__day__temperature">
          <span className="weather-forecast__day__temperature--max">
            {day.max}&deg;
          </span>
          <span className="weather-forecast__day__temperature__separator">/</span>
          <span className="weather-forecast__day__temperature--min">
            {day.min}&deg;
          </span>
        </div>
      </div>
    ))}
  </div>
);

export default WeatherForecast;
