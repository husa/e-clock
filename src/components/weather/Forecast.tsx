import './forecast.scss';

import type { WeatherResponseForecast } from '../../types';
import lang from '../../services/lang';
import Icon from './Icon';

const ONE_DAY = 24 * 60 * 60 * 1000;

export const daysBetween = (day1: string, day2: string) => {
  const d1 = new Date(day1);
  const d2 = new Date(day2);
  return ((d2.getTime() - d1.getTime()) / ONE_DAY) | 0;
};

export const getDay = (date: string) => {
  if (daysBetween(new Date().toDateString(), date) === 0) return lang.t('i18nToday');
  if (daysBetween(new Date().toDateString(), date) === 1) return lang.t('i18nTomorrow');
  return lang.t(`i18nDay${new Date(date).getDay()}` as 'i18nDay0');
};

type Props = {
  data: WeatherResponseForecast;
};

const WeatherForecast = (props: Props) => (
  <div className="weather-forecast">
    {props.data.map((day) => (
      <div className="weather-forecast__day" key={day.date}>
        <div className="weather-forecast__day__name">{getDay(day.date)}</div>
        <Icon className="weather-forecast__day__icon" code={day.code} title={day.text} />
        <div className="weather-forecast__day__temperature">
          <span className="weather-forecast__day__temperature--max">{day.max}&deg;</span>
          <span className="weather-forecast__day__temperature__separator">/</span>
          <span className="weather-forecast__day__temperature--min">{day.min}&deg;</span>
        </div>
      </div>
    ))}
  </div>
);

export default WeatherForecast;
