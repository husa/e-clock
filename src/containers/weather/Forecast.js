import {connect} from 'react-redux';

import {
  selectWeatherForecast
} from '../../selectors';

import WeatherForecast from '../../components/weather/Forecast';

const mapStateToProps = state => ({
  data: selectWeatherForecast(state)
});

export default connect(
  mapStateToProps
)(WeatherForecast);
