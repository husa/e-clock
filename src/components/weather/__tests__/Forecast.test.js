import renderer from 'react-test-renderer';

import Forecast, { daysBetween, getDay } from '../Forecast';

describe('Components > weather > Forecast', () => {
  const mockToday = (today, fn) => {
    /* global global */
    const OriginalDate = Date;
    // mock Date
    global.Date = jest.fn().mockImplementation((a) => new OriginalDate(a ? a : today));
    global.Date.now = OriginalDate.now;
    // run function
    fn();
    // restore mocked Date
    global.Date = OriginalDate;
  };

  it('should match snapshot', () => {
    const testData = [
      { code: '1', date: '30 Jan 2018', day: 'Tue', max: 10, min: 5, text: 'test1' },
      { code: '3', date: '18 Jan 2018', day: 'Wed', max: 9, min: 3, text: 'test2' },
      { code: '5', date: '19 Jan 2018', day: 'Thu', max: -100, min: 500, text: 'test3' },
      { code: '2', date: '2 Feb 2018', day: 'Fri', max: 100, min: -273, text: 'test4' },
      { code: '18', date: '3 Feb 2018', day: 'Sat', max: 10, min: 5, text: 'test5' },
    ];
    mockToday('18 Jan 2018', () => {
      const component = renderer.create(<Forecast data={testData} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('daysBetween', () => {
    it('should return days between two dates', () => {
      expect(daysBetween('18 Jan 2018', '18 Jan 2018')).toEqual(0);
      expect(daysBetween('18 Jan 2018', '19 Jan 2018')).toEqual(1);
      expect(daysBetween('18 Jan 2018', '20 Jan 2018')).toEqual(2);
      expect(daysBetween('18 Jan 2018', '17 Jan 2018')).toEqual(-1);
      expect(daysBetween('18 Jan 2018', '18 Feb 2018')).toEqual(31);
      expect(daysBetween('18 Feb 2018', '18 Mar 2018')).toEqual(28);
    });
  });

  describe('getDay', () => {
    it('should return day name for date', () => {
      mockToday('18 Jan 2018', () => {
        expect(getDay('17 Jan 2018')).toEqual('Day3');
        expect(getDay('18 Jan 2018')).toEqual('Today');
        expect(getDay('19 Jan 2018')).toEqual('Tomorrow');
        expect(getDay('20 Jan 2018')).toEqual('Day6');
        expect(getDay('21 Jan 2018')).toEqual('Day0');
      });
    });
  });
});
