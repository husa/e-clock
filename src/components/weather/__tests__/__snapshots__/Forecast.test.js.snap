// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components > weather > Forecast should match snapshot 1`] = `
<div
  className="weather-forecast"
>
  <div
    className="weather-forecast__day"
  >
    <div
      className="weather-forecast__day__name"
    >
      Day2
    </div>
    <img
      alt="test1"
      className="weather-icon weather-forecast__day__icon"
      src="/img/weather-icons/weather_1.png"
      title="test1"
    />
    <div
      className="weather-forecast__day__temperature"
    >
      <span
        className="weather-forecast__day__temperature--max"
      >
        10
        °
      </span>
      <span
        className="weather-forecast__day__temperature__separator"
      >
        /
      </span>
      <span
        className="weather-forecast__day__temperature--min"
      >
        5
        °
      </span>
    </div>
  </div>
  <div
    className="weather-forecast__day"
  >
    <div
      className="weather-forecast__day__name"
    >
      Today
    </div>
    <img
      alt="test2"
      className="weather-icon weather-forecast__day__icon"
      src="/img/weather-icons/weather_3.png"
      title="test2"
    />
    <div
      className="weather-forecast__day__temperature"
    >
      <span
        className="weather-forecast__day__temperature--max"
      >
        9
        °
      </span>
      <span
        className="weather-forecast__day__temperature__separator"
      >
        /
      </span>
      <span
        className="weather-forecast__day__temperature--min"
      >
        3
        °
      </span>
    </div>
  </div>
  <div
    className="weather-forecast__day"
  >
    <div
      className="weather-forecast__day__name"
    >
      Tomorrow
    </div>
    <img
      alt="test3"
      className="weather-icon weather-forecast__day__icon"
      src="/img/weather-icons/weather_5.png"
      title="test3"
    />
    <div
      className="weather-forecast__day__temperature"
    >
      <span
        className="weather-forecast__day__temperature--max"
      >
        -100
        °
      </span>
      <span
        className="weather-forecast__day__temperature__separator"
      >
        /
      </span>
      <span
        className="weather-forecast__day__temperature--min"
      >
        500
        °
      </span>
    </div>
  </div>
  <div
    className="weather-forecast__day"
  >
    <div
      className="weather-forecast__day__name"
    >
      Day5
    </div>
    <img
      alt="test4"
      className="weather-icon weather-forecast__day__icon"
      src="/img/weather-icons/weather_2.png"
      title="test4"
    />
    <div
      className="weather-forecast__day__temperature"
    >
      <span
        className="weather-forecast__day__temperature--max"
      >
        100
        °
      </span>
      <span
        className="weather-forecast__day__temperature__separator"
      >
        /
      </span>
      <span
        className="weather-forecast__day__temperature--min"
      >
        -273
        °
      </span>
    </div>
  </div>
  <div
    className="weather-forecast__day"
  >
    <div
      className="weather-forecast__day__name"
    >
      Day6
    </div>
    <img
      alt="test5"
      className="weather-icon weather-forecast__day__icon"
      src="/img/weather-icons/weather_18.png"
      title="test5"
    />
    <div
      className="weather-forecast__day__temperature"
    >
      <span
        className="weather-forecast__day__temperature--max"
      >
        10
        °
      </span>
      <span
        className="weather-forecast__day__temperature__separator"
      >
        /
      </span>
      <span
        className="weather-forecast__day__temperature--min"
      >
        5
        °
      </span>
    </div>
  </div>
</div>
`;
